modules/index/src/test/java/org/apache/ignite/internal/index/IndexAvailabilityControllerTest.java

```diff
@@ -0,0 +1,405 @@
+        vaultManager = new VaultManager(new InMemoryVaultService());
+
+        metaStorageManager = StandaloneMetaStorageManager.create(vaultManager);
+
+        catalogManager = CatalogTestUtils.createTestCatalogManager(NODE_NAME, clock, metaStorageManager);
+
+        indexBuilder = new IndexBuilder(NODE_NAME, 1, mock(ReplicaService.class, invocation -> completedFuture(null)));
+
+        indexAvailabilityController = new IndexAvailabilityController(catalogManager, metaStorageManager, indexBuilder);
+
+        Stream.of(vaultManager, metaStorageManager, catalogManager).forEach(IgniteComponent::start);
+
+        assertThat(metaStorageManager.deployWatches(), willCompleteSuccessfully());
+
+        CatalogZoneDescriptor zoneDescriptor = catalogManager.zone(DEFAULT_ZONE_NAME, clock.nowLong());
+
+        assertNotNull(zoneDescriptor);
+
+        partitions = zoneDescriptor.partitions();
+
+        assertThat(partitions, greaterThan(4));
+
+        TableTestUtils.createTable(
+                catalogManager,
+                DEFAULT_SCHEMA_NAME,
+                DEFAULT_ZONE_NAME,
+                TABLE_NAME,
+                List.of(ColumnParams.builder().name(COLUMN_NAME).type(INT32).build()),
+                List.of(COLUMN_NAME)
+        );
+    }
+
+    @AfterEach
+    void tearDown() throws Exception {
+        IgniteUtils.closeAll(
+                indexAvailabilityController == null ? null : indexAvailabilityController::close,
+                indexBuilder == null ? null : indexBuilder::close,
+                catalogManager == null ? null : catalogManager::stop,
+                metaStorageManager == null ? null : metaStorageManager::stop,
+                vaultManager == null ? null : vaultManager::stop
+        );
+    }
+
+    @Test
+    void testMetastoreKeysAfterIndexCreate() throws Exception {
+        createIndex(INDEX_NAME);
+
+        int indexId = indexId(INDEX_NAME);
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyExists(indexId);
+
+        for (int partitionId = 0; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyExists(indexId, partitionId);
+        }
+    }
+
+    @Test
+    void testMetastoreKeysAfterIndexCreateForOnlyOnePartition() throws Exception {
+        changePartitionCountInCatalog(1);
+
+        testMetastoreKeysAfterIndexCreate();
+    }
+
+    @Test
+    void testMetastoreKeysAfterFinishBuildIndexForOnePartition() throws Exception {
+        createIndex(INDEX_NAME);
+
+        int indexId = indexId(INDEX_NAME);
+
+        finishBuildingIndexForPartition(indexId, 0);
+
+        awaitActualMetastoreRevision();
+
+        assertStartBuildIndexKeyExists(indexId);
+
+        assertPartitionBuildIndexKeyAbsent(indexId, 0);
+
+        for (int partitionId = 1; partitionId < partitions; partitionId++) {
+            assertPartitionBuildIndexKeyExists(indexId, partitionId);
+        }
+
+        assertTrue(indexDescriptor(INDEX_NAME).writeOnly());
```

Comment:
`awaitActualMetastoreRevision()` seems to wait till the local applied revision reaches the current global MetaStorage revision. But switching the index state in the controller happens asynchronously with respect to the watch thread. Is there a guarantee that at line 177 we'll see the new index descriptor version? Is it possible that we only wait for some revision N in `awaitActualMetastoreRevision`, while the index will be updated in revision N+1?

Reply:
As far as I understand, if we made N global changes and waited for them, and in parallel did not change anything else, then we are guaranteed to wait for all the changes to be applied in the metastore.

Reply:
But doesn't the code under test makes a parallel change?

Reply:
Nope.